CARDS-2021: PREMs: Automatically delete submitted survey responses older than one year

veronikaslc opened the pull request.

veronikaslc committed:
CARDS-2021: PREMs: Automatically delete submitted survey responses older than one year

sdumitriu commented on prems-resources/backend/src/main/java/io/uhndata/cards/prems/patients/internal/SubmittedFormsCleanupScheduler.java:
```diff
@@ -0,0 +1,93 @@
+ * @version $Id$
+ * @since 0.9.6
+ */
+@Designate(ocd = SubmittedFormsCleanupScheduler.Config.class, factory = true)
```
This doesn't need more than one instance running, so it shouldn't be using a factory.

Or, alternatively, add another setting in the `Config` below for the questionnaires to delete, so we could only have different deletion policies for different data.

veronikaslc committed:
CARDS-2021: PREMs: Automatically delete submitted survey responses older than one year
Fixed tag

marta- commented on prems-resources/backend/src/main/java/io/uhndata/cards/prems/patients/internal/SubmittedFormsCleanupTask.java:
```diff
@@ -0,0 +1,123 @@
+
+            // Query:
+            final Iterator<Resource> resources = resolver.findResources(String.format(
+                // select the data forms
+                "select distinct dataForm.*"
+                    + "  from [cards:Form] as dataForm"
+                    // belonging to a visit
+                    + "  inner join [cards:Form] as visitInformation on visitInformation.subject = dataForm.subject"
+                    + "    inner join [cards:Answer] as submitted on isdescendantnode(submitted, visitInformation)"
+                    + " where"
+                    // link to the correct Visit Information questionnaire
+                    + "  visitInformation.questionnaire = '%1$s'"
+                    // the data form was last modified by the patient before the allowed timeframe
+                    + "  and dataForm.[jcr:lastModified] < '%2$s'"
+                    // the visit is submitted
+                    + "  and submitted.question = '%3$s'"
+                    + "  and submitted.value = 1"
+                    // exclude the Visit Information form itself
+                    + "  and dataForm.questionnaire <> '%1$s'",
```
This query can probably be made more efficient by using looking for the `SUBMITTED` status flag and replacing the use of `isdescendantnode` with the `form` property (if still necessary).

veronikaslc replied:
Not sure I got the optimisation of `isdescendantnode` part. Also did you mean the status flag on the form? May be the  proposed change would be helpful

sdumitriu reviewed (commented):
This should be moved in the `patient-portal` module.

veronikaslc committed:
CARDS-2021: PREMs: Automatically delete submitted survey responses older than one year
Moved files to the patient-portal module; Removed factory usage from the scheduler

veronikaslc commented:
Rebased and addressed most of the comments except query optimisation.
Not tested yet, needs instructions.
